core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -207,68 +214,96 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
                nil -> makeStage(prefetchKeysIfNeededAndInvokeNext(ctx, command, keys, true)).toCompletableFuture()));
       }
       if (transferedKeys != null) {
-         return asyncInvokeNext(ctx, command, retrieveRemoteValues(ctx, transferedKeys).toCompletableFuture());
+         return asyncInvokeNext(ctx, command, retrieveRemoteValues(ctx, transferedKeys, command.getTopologyId()).toCompletableFuture());
       } else {
          return invokeNext(ctx, command);
       }
    }
 
-   private InvocationStage retrieveSingleRemoteValue(InvocationContext ctx, Object key) {
+   private InvocationStage lookupLocalAndRetrieveRemote(InvocationContext ctx, Object key, int topologyId) {
+      if (trace) {
+         log.tracef("Locally prefetching entry for key %s", key);
+      }
       GetCacheEntryCommand getCacheEntryCommand = commandsFactory.buildGetCacheEntryCommand(key, EnumUtil.bitSetOf(Flag.CACHE_MODE_LOCAL));
-      return makeStage(invokeNextThenApply(ctx, getCacheEntryCommand, (ctx1, command1, rv) -> {
-         CacheEntry entry = (CacheEntry) rv;
-         Metadata metadata = entry != null ? entry.getMetadata() : null;
-         CompletableFuture<InternalCacheValue> future;
-         if (metadata != null && metadata.version() != null && svm.isVersionActual(keyPartitioner.getSegment(key), metadata.version())){
-            return null;
-         } else if ((metadata instanceof RemoteMetadata)) {
-            Address backup = ((RemoteMetadata) metadata).getAddress();
-            future = retrieveRemoteValue(Collections.singleton(backup), key);
-         } else {
-            future = retrieveRemoteValue(null, key);
+      getCacheEntryCommand.setTopologyId(topologyId);
+      return makeStage(invokeNextThenApply(ctx, getCacheEntryCommand, handleLocallyLookedUpEntry));
+   }
+
+   private Object handleLocallyLookedUpEntry(InvocationContext ctx1, VisitableCommand command1, Object rv) {
+      GetCacheEntryCommand cmd = (GetCacheEntryCommand) command1;
+      if (trace) {
+         log.tracef("Locally prefetched entry %s", rv);
+      }
+      CacheEntry entry = (CacheEntry) rv;
+      Metadata metadata = entry != null ? entry.getMetadata() : null;
+      CompletableFuture<InternalCacheValue> future;
+      int segment = keyPartitioner.getSegment(cmd.getKey());
+      if (metadata != null && metadata.version() != null && svm.isVersionActual(segment, metadata.version())){
+         return null;
+      } else if ((metadata instanceof RemoteMetadata) && svm.getSegmentState(segment) == ScatteredVersionManager.SegmentState.VALUE_TRANSFER) {
+         // The RemoteMetadata is valid only during value transfer - in blocked state there shouldn't be any such
+         // entry and during key transfer we could see metadata pointing to a node with outdated information.
```

Comment:
Maybe you could add some javadoc to `RemoteMetadata` to explain what `getAddress()` means during each phase?